### PR TITLE
AArch64: Add UEFI based direct kernel boot support

### PR DESCRIPTION
**Pull request by jongwu**
UEFI boot on AArch64 has not completed, see issue #5981.
This PR tries to enable all of the uncompleted tasks on CH side.

- [x] Boot from RHF using "-firmware";
- [x] boot from RHF using "-frimware" and start kernel specified by "-kernel", need support in RHF, see [PR](https://github.com/cloud-hypervisor/rust-hypervisor-firmware/pull/297)
- [x] boot from EDK2 using "-firmware" and start kernel specified by "-kernel", need support in EDK2, patch in review [ml](https://edk2.groups.io/g/devel/message/93882?p=%2C%2C%2C20%2C0%2C0%2C0%3A%3Acreated%2C0%2Cjianyong%2C20%2C2%2C40%2C93715215)

### CODE REVIEW

File: vmm/src/vm.rs [hunk 1]
```diff
@@ -920,6 +940,42 @@ impl Vm {
...
+            .map_err(|_| Error::SeekKernelStart)?;
+
+        let kernel_addr = if guest_addr.0 == 0 {
+            arch::initramfs_load_addr(guest_mem, kernel_size).map_err(|_| Error::KernelLoadMem)?
+        } else {
+            guest_addr.0
+        };
+
+        guest_mem
+            .read_exact_from(GuestAddress(kernel_addr), &mut kernel_image, kernel_size)
```
Comment (likebreath) [inline vmm/src/vm.rs:970]: If this is why you have to use `#[allow(deprecated)]`, you can use its alternative `read_exact_volatile_from`. And this commit is a good reference: https://github.com/cloud-hypervisor/cloud-hypervisor/commit/d4892f41b3124e777724ed872fb9997b8a47c1fb

File: vmm/src/vm.rs [hunk 2]
```diff
@@ -929,39 +985,45 @@ impl Vm {
         Ok(())
     }
 
+    // Load image as pe file first, if fails, load it as plain file.
+    #[cfg(target_arch = "aarch64")]
+    fn load_pe_or(
```
Comment (michael2012z) [inline vmm/src/vm.rs:990]: I think we don't need to add yet another function. See the comment on `load_payload()`.

And some question regarding this code changes: 
- If I assign a kernel binary to `--firmware`, does the guest boots from that kernel? If yes, that's not expected.
- If I assign a kernel to `--firmware` and the same kernel to `--kernel`, does the guest boots?

File: vmm/src/vm.rs [hunk 3]
```diff
@@ -1041,6 +1103,11 @@ impl Vm {
                 let kernel = File::open(kernel).map_err(Error::KernelFile)?;
                 Self::load_kernel(None, Some(kernel), memory_manager)
             }
+            (Some(firmware), Some(kernel)) => {
+                let firmware = File::open(firmware).map_err(Error::FirmwareFile)?;
+                let kernel = File::open(kernel).map_err(Error::KernelFile)?;
+                Self::load_kernel(Some(firmware), Some(kernel), memory_manager)
+            }
```
Comment (michael2012z) [inline vmm/src/vm.rs:1110]: Can we achieve the booting with both firmware and kernel simply in this way?
- Keep the implementation of `load_kernel`
- Invoke `Self::load_kernel(None, Some(kernel), memory_manager);` to load the kernel at first and invoke `Self::load_kernel(Some(firmware), None, memory_manager)` and return the `EntryPoint` of the firmware

**Review by michael2012z — CHANGES_REQUESTED** [review]

File: vmm/src/vm.rs [hunk 4]
```diff
@@ -1151,6 +1218,26 @@ impl Vm {
             None => None,
         };
 
+        let kernel_addr = match initramfs_config {
+            Some(ref initrd) => initrd.address.0 - arch::aarch64::layout::KERNEL_IMAGE_MAX_SIZE,
+            None => 0,
+        };
```
[the remark this answers is not in the document]
Comment (jongwu) [reply vmm/src/vm.rs:1224]: kernel image here servers as data passed to firmware to load. So, it should not be subjected to this definition. Just like initramfs does.

**Comment by michael2012z** [comment]
Hi, @jongwu 

Here are my concerns on the code changes:
1. `load_kernel()` now becomes a wrapper of `load_pe_or()`. It is not necessary to introduce yet another function. 
2. The function name `load_pe_or` is confusing.
3. Kernel loading is done in 2 different functions: `load_payload()` and `load_kernel_to_mem()`
4. For `KernelInfo`, it's better to avoid introducing a AArch64 specific struct. And it's very similar to `InitramfsConfig`.

For 1, 2 and 3, is it better to improve in this way?

- Rename `load_pe_or()` to `load_booting_binary()`
- Pass the binary and target address to `load_booting_binary()`, the binary could be a kernel or a firmware. Keep the current logic in it: Try loading as PE at first, if failed try loading as firmware.
```
fn load_booting_binary(
        binary: Option<File>,
        address: GuestAddress,
        memory_manager: Arc<Mutex<MemoryManager>>,
    ) -> Result<EntryPoint> ...
```

- Remove `load_kernel()`
- Invoke `load_booting_binary()` from `load_payload()`, the pseudo code is like:
```
    match (&payload.firmware, &payload.kernel) {
            (Some(firmware), None) => {
                Self::load_booting_binary(Some(firmware), arch::layout::UEFI_START, memory_manager)
            }
            (None, Some(kernel)) => {
                Self::load_booting_binary(Some(kernel), arch::layout::KERNEL_START, memory_manager)
            }
            (Some(firmware), Some(kernel)) => {
            	let kernel_addr = Calculate the target kernel address that the firmware needs;
            	_ = Self::load_booting_binary(Some(kernel), kernel_addr, memory_manager);
                Self::load_booting_binary(Some(firmware), arch::layout::UEFI_START, memory_manager)
            }
```
- In `system_configure()`, calculate the kernel target address and size for FDT, but do not load it into guest memory.

**Comment by jongwu** [comment]
Thanks @michael2012z -, Great refactor.

I find a issue. PE loader will load kernel to some other place than the address passed to it and will do lots of check that no need for my case.  What I need is that load kernel image as a overall data into memory without unpacking it just like loading initramfs.
So, I need introduce a new loader or refact then reuse "load_initramfs". Which do you perfer?

**Comment by michael2012z** [comment]
> I find a issue. PE loader will load kernel to some other place than the address passed to it and will do lots of check that no need for my case. What I need is that load kernel image as a overall data into memory without unpacking it just like loading initramfs. So, I need introduce a new loader or refact then reuse "load_initramfs". Which do you perfer?

In that case, I prefer to refactor `load_initramfs()` and `InitramfsConfig` for general usage covering initramfs and kernel image. In this way, we also resolved the problem of introducing a new KernelConfig, right?

**Comment by jongwu** [comment]
> In that case, I prefer to refactor `load_initramfs()` and `InitramfsConfig` for general usage covering initramfs and kernel image. In this way, we also resolved the problem of introducing a new KernelConfig, right?

Make sense. Let me try this way.

File: vmm/src/vm.rs [hunk 5]
```diff
@@ -883,25 +886,36 @@ impl Vm {
...
 
-        let address =
+        let top_addr =
             arch::initramfs_load_addr(guest_mem, size).map_err(|_| Error::InitramfsLoad)?;
-        let address = GuestAddress(address);
+        #[cfg(target_arch = "aarch64")]
+        let offset = arch::aarch64::layout::KERNEL_IMAGE_MAX_SIZE;
+        #[cfg(target_arch = "x86_64")]
+        let offset = 0;
+        let address = if is_top { top_addr } else { top_addr - offset };
```
Comment (michael2012z) [inline vmm/src/vm.rs:910]: This logic is only for AArch64, X86_64 has nothing with `top_addr` and `is_top`. Is it better if you calculate the address on the caller side and keep the logic of this function simple? Specifically:

1. Replace the parameter `is_top: bool,` with `addr: u64`
2. In this function, simply load the file to the specified `addr`
3. On X86_64, pass `arch::initramfs_load_addr()` when calling this function
4. On AArch64, apply your logic to determine the address in `configure_system()` before calling this function

File: arch/src/lib.rs [hunk 6]
```diff
@@ -117,11 +117,11 @@ pub struct NumaNode {
 
 pub type NumaNodes = BTreeMap<u32, NumaNode>;
 
-/// Type for passing information about the initramfs in the guest memory.
-pub struct InitramfsConfig {
-    /// Load address of initramfs in guest memory
+/// Type for passing information about the initramfs/kernel in the guest memory.
+pub struct BootFileConfig {
```
Comment (michael2012z) [inline arch/src/lib.rs:121]: The name `BootFileConfig` sounds too generic, but I am not good at naming.

Any better idea for the name?

**Comment by likebreath** [comment]
Closing due to no activities. Please re-open if needed. Thank you.

**Pull request closed without merging**